witness/internal/witness/db.go

```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package witness
+
+import (
+	"database/sql"
+	"fmt"
+)
+
+// Database simply gets and puts things into persistent storage.
+type Database struct {
+	db *sql.DB
+}
+
+// NewDatabase creates a new database, initializing it if needed.
+func NewDatabase(db *sql.DB) (*Database, error) {
+	d := &Database{
+		db: db,
+	}
+	return d, d.init()
+}
+
+func (d *Database) init() error {
+	_, err := d.db.Exec("CREATE TABLE IF NOT EXISTS chkpts (key BLOB PRIMARY KEY, size INT, data BLOB)")
+	return err
+}
+
+// GetLatest reads the latest checkpoint written to the DB for a given log.
+func (d *Database) GetLatest(logPK string) (*Chkpt, error) {
+	var maxChkpt *Chkpt
+	row := d.db.QueryRow("SELECT data FROM chkpts WHERE key = ? ORDER BY size DESC LIMIT 1", logPK)
+	if err := row.Err(); err != nil {
+		return nil, err
+	}
+	if err := row.Scan(&maxChkpt); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, fmt.Errorf("Unknown public key: %q", err)
+		}
+		return nil, err
+	}
+	return maxChkpt, nil
+}
+
+// SetCheckpoint writes the checkpoint to the DB for a given logId.
+func (d *Database) SetCheckpoint(logPK string, c Chkpt) error {
+	_, err := d.db.Exec("INSERT OR IGNORE INTO chkpts (key, size, data) VALUES (?, ?, ?)", logPK, c.Parsed.Size, c)
```

Comment:
Does this work?! `insert or ignore` should refuse to overwrite anything else. This would be OK if the DB key for the table was composite on (key, size), but as the table is set up with the key being only the log id, this looks to me like the first checkpoint set would be the only checkpoint ever set for the log.

Reply:
Suggest adding a test for this. With sqlite you can do this with `:memory:`; search the repo for that and you'll find examples.

Reply:
It did not work, no :).  The primary key is now composite though so it does (see the new test file to prove it).